samples/bluetooth/conn_time_sync/src/controller_time_nrf53_app.c

```diff
@@ -0,0 +1,250 @@
+
+	nrfx_gppi_channels_enable(BIT(ppi_chan_timer_clear_on_rtc_tick));
+
+	nrfx_timer_enable(&app_timer_instance);
+
+	return 0;
+}
+
+/** Configure the TIMER and RTC in such a way that microsecond accurate timing can be achieved.
+ *
+ * To get microsecond accurate toggling, we need to combine both the RTC and TIMER peripheral.
+ * That is, both a RTC CC value and TIMER CC value needs to be set.
+ * We should only trigger an EGU task when the TIMER CC value
+ * matches after the RTC CC value matched.
+ * This is achieved using a PPI group. The group is enabled when the RTC CC matches and disabled
+ * again then the TIMER CC matches.
+ */
+int config_egu_trigger_on_rtc_and_timer_match(void)
+{
+	uint8_t ppi_chan_on_rtc_match;
+	uint8_t ppi_chan_on_timer_match;
+
+	if (nrfx_gppi_channel_alloc(&ppi_chan_on_rtc_match) != NRFX_SUCCESS) {
+		printk("Failed allocating for RTC match\n");
+		return -ENOMEM;
+	}
+
+	if (nrfx_gppi_channel_alloc(&ppi_chan_on_timer_match) != NRFX_SUCCESS) {
+		printk("Failed allocating for TIMER match\n");
+		return -ENOMEM;
+	}
+
+	nrfx_gppi_group_clear(NRFX_GPPI_CHANNEL_GROUP0);
+	nrfx_gppi_group_disable(NRFX_GPPI_CHANNEL_GROUP0);
+	nrfx_gppi_channels_include_in_group(BIT(ppi_chan_on_timer_match),
+					    NRFX_GPPI_CHANNEL_GROUP0);
+
+	nrfx_gppi_channel_endpoints_setup(ppi_chan_on_rtc_match,
+					  nrfx_rtc_event_address_get(&app_rtc_instance,
+								     NRF_RTC_EVENT_COMPARE_0),
+					  nrfx_gppi_task_address_get(NRFX_GPPI_TASK_CHG0_EN));
+	nrfx_gppi_channel_endpoints_setup(ppi_chan_on_timer_match,
+					  nrfx_timer_event_address_get(&app_timer_instance,
+								       NRF_TIMER_EVENT_COMPARE0),
+					  nrfx_gppi_task_address_get(NRFX_GPPI_TASK_CHG0_DIS));
+	nrfx_gppi_fork_endpoint_setup(ppi_chan_on_timer_match,
+				      nrf_egu_task_address_get(NRF_EGU0, NRF_EGU_TASK_TRIGGER0));
+
+	nrfx_gppi_channels_enable(BIT(ppi_chan_on_rtc_match));
+
+	return 0;
+}
+
+int controller_time_init(void)
+{
+	int ret;
+
+	ret = rtc_config();
+	if (ret) {
+		return ret;
+	}
+
+	ret = timer_config();
+	if (ret) {
+		return ret;
+	}
+
+	return config_egu_trigger_on_rtc_and_timer_match();
+}
+
+static uint64_t rtc_ticks_to_us(uint32_t rtc_ticks)
+{
+	const uint64_t rtc_ticks_in_femto_units = 30517578125UL;
+
+	return (rtc_ticks * rtc_ticks_in_femto_units) / 1000000000UL;
+}
+
+static uint32_t us_to_rtc_ticks(uint32_t timestamp_us)
+{
+	const uint64_t rtc_ticks_in_femto_units = 30517578125UL;
+
+	return ((uint64_t)(timestamp_us) * 1000000000UL) / rtc_ticks_in_femto_units;
+}
+
+uint64_t controller_time_us_get(void)
+{
+	/* Simply use the RTC time here.
+	 * It is possible to combine RTC and TIMER information here to get a more accurate
+	 * timestamp. That requires setting up a PPI channel to capture both values simultaneously.
+	 */
+
+	const uint64_t rtc_overflow_time_us = 512000000UL;
+
+	int key = irq_lock();
+
+	uint32_t rtc_ticks = nrf_rtc_counter_get(app_rtc_instance.p_reg);
+	uint32_t rtc_overflows = num_rtc_overflows;
```

Comment:
I believe we have the reverse problem now. The RTC may have overflowed after we disabled the interrupt on line 210, but this is not reflected in `num_rtc_overflows`